sql: fix limitations in, and improve, GROUP BY and ORDER BY

knz opened the pull request:
Informs #15057.

knz committed:
sql: expose the grouping columns in the output of EXPLAIN

knz committed:
sql: simplify the code for GROUP BY, fix a limitation

Prior to this patch, star expansion in a GROUP BY clause would not
"remember" the results of the expansion when deciding whether or not a
SELECT render was valid.

For example:

```
> SELECT k FROM kv GROUP BY kv.*
pq: column "k" must appear in the GROUP BY clause or be used in an aggregate function
```

This limitation is spurious since `kv.*` implies `k` so `k` does
actually "appear" in the GROUP BY clause.

This patch fixes that limitation.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17888)
<!-- Reviewable:end -->

knz committed:
sql: flatten tuples in GROUP BY and ORDER BY expressions

Prior to this patch, a query like `SELECT v FROM kv GROUP BY (v, v)`
would be rejected with the error message "`v does not appear in GROUP
BY`". Also `SELECT * FROM kv ORDER BY (k, v)` would cause a spurious
extra sorting stage even though the table is already in the
specified order.

This patch resolves these limitations by "opening the tuples" when
computing GROUP BY and ORDER BY.

Note that this does *not* optimize sorting when there are intermediate
tuples generated; for example:

```
SELECT v FROM (SELECT v, (k, v) FROM kv) ORDER BY 2
```

This query also adds an extraneous sorting stage, even though the
inner query is guaranteed to deliver the tuples `(k, v)` in order.

(Incidentally, tuples cause issues with distributed queries. Both of
these issues are discussed further in #15057.)

RaduBerinde commented:
LGTM, nice fixes.
